### PR TITLE
Trigger caching headers finalization on write

### PR DESCRIPTION
**Pull request by JunTaoLuo**
#92 

- [x] Add tests for Write and WriteAsync
- [x] Test with compression middleware

### CODE REVIEW

File: src/Microsoft.AspNetCore.ResponseCaching/Internal/MemoryResponseCache.cs [hunk 1]
```diff
@@ -22,34 +22,40 @@ public MemoryResponseCache(IMemoryCache cache)
...
+
+        public void Set(string key, IResponseCacheEntry entry, TimeSpan validFor)
         {
             var cachedResponse = entry as CachedResponse;
             if (cachedResponse != null)
             {
                 var segmentStream = new SegmentWriteStream(StreamUtilities.BodySegmentSize);
-                await cachedResponse.Body.CopyToAsync(segmentStream);
+                // We are guaranteed to have a SegmentReadStream.CopyToAsync(SegmentWriteStream)
+                cachedResponse.Body.CopyToAsync(segmentStream).Wait();
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.ResponseCaching/Internal/MemoryResponseCache.cs]: CopyTo

File: src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingMiddleware.cs [hunk 2]
```diff
@@ -219,10 +214,11 @@ public class ResponseCachingMiddleware
             return false;
         }
 
-        internal async Task FinalizeCacheHeadersAsync(ResponseCachingContext context)
+        private bool FinalizeCacheHeadersShouldStoreVaryByEntry(ResponseCachingContext context)
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingMiddleware.cs]: This name is a bit silly, a doc comment would be sufficient to explain the return parameter.

File: src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingMiddleware.cs [hunk 3]
```diff
@@ -327,19 +343,33 @@ public class ResponseCachingMiddleware
             }
         }
 
-        internal Task OnResponseStartingAsync(ResponseCachingContext context)
+        private bool OnResponseStartingShouldFinalizeHeaders(ResponseCachingContext context)
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingMiddleware.cs]: StartResponse

File: src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingMiddleware.cs [hunk 4]
```diff
@@ -355,7 +385,12 @@ internal void ShimResponseStream(ResponseCachingContext context)
         {
             // Shim response stream
             context.OriginalResponseStream = context.HttpContext.Response.Body;
-            context.ResponseCachingStream = new ResponseCachingStream(context.OriginalResponseStream, _options.MaximumBodySize, StreamUtilities.BodySegmentSize);
+            context.ResponseCachingStream = new ResponseCachingStream(
+                context.OriginalResponseStream,
+                _options.MaximumBodySize,
+                StreamUtilities.BodySegmentSize,
+                () => OnResponseStarting(context),
+                async () => await OnResponseStartingAsync(context));
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingMiddleware.cs]: no async await needed?

**Comment by JunTaoLuo** [comment]
:up::date:

**Pull request merged**